generate Nodejs from autorest

DeepakRajendranMsft opened the pull request.

azurecla commented:
Hi __@DeepakRajendranMsft__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

amarzavery commented:
@DeepakRajendranMsft - The CI is failing due to jshint issues - https://travis-ci.org/Azure/azure-sdk-for-node/jobs/93226633 

This character in swagger spec Asterix â€œ*â€ is wacky.  Looks like double quotes are in MS Word format “*”. They should rather be like this: "*"

```
683<jslint>

684	<file name="lib/services/networkManagement2/lib/models/securityRule.js">

685		<issue line="36" char="55" reason="This character may get silently deleted by one or more browsers." evidence=" * Integer or range between 0 and 65535. Asterix â€œ*â€ can also be used to" severity="W" />

686		<issue line="40" char="62" reason="This character may get silently deleted by one or more browsers." evidence=" * Range. Integer or range between 0 and 65535. Asterix â€œ*â€ can also be" severity="W" />

687		<issue line="44" char="42" reason="This character may get silently deleted by one or more browsers." evidence=" * CIDR or source IP range. Asterix â€œ*â€ can also be used to match all" severity="W" />

688		<issue line="50" char="50" reason="This character may get silently deleted by one or more browsers." evidence=" * prefix. CIDR or source IP range. Asterix â€œ*â€ can also be used to match" severity="W" />

689		<issue line="55" char="33" reason="This character may get silently deleted by one or more browsers." evidence=" * Possible values are â€œAllowâ€ and â€œDenyâ€. Possible values for this" severity="W" />

690	</file>

691	<file name="lib/services/networkManagement2/lib/operations/securityRules.js">

692		<issue line="406" char="9" reason="This character may get silently deleted by one or more browsers." evidence=" * â€œ*â€ can also be used to match all ports." severity="W" />

693		<issue line="410" char="9" reason="This character may get silently deleted by one or more browsers." evidence=" * â€œ*â€ can also be used to match all ports." severity="W" />

694		<issue line="413" char="65" reason="This character may get silently deleted by one or more browsers." evidence=" * source address prefix. CIDR or source IP range. Asterix â€œ*â€ can also" severity="W" />

695		<issue line="420" char="75" reason="This character may get silently deleted by one or more browsers." evidence=" * sets destination address prefix. CIDR or source IP range. Asterix â€œ*â€" severity="W" />

696		<issue line="426" char="63" reason="This character may get silently deleted by one or more browsers." evidence=" * traffic is allowed or denied. Possible values are â€œAllowâ€ and" severity="W" />

697		<issue line="427" char="12" reason="This character may get silently deleted by one or more browsers." evidence=" * â€œDenyâ€. Possible values for this property include: &apos;Allow&apos;, &apos;Deny&apos;." severity="W" />

698		<issue line="546" char="9" reason="This character may get silently deleted by one or more browsers." evidence=" * â€œ*â€ can also be used to match all ports." severity="W" />

699		<issue line="550" char="9" reason="This character may get silently deleted by one or more browsers." evidence=" * â€œ*â€ can also be used to match all ports." severity="W" />

700		<issue line="553" char="65" reason="This character may get silently deleted by one or more browsers." evidence=" * source address prefix. CIDR or source IP range. Asterix â€œ*â€ can also" severity="W" />

701		<issue line="560" char="75" reason="This character may get silently deleted by one or more browsers." evidence=" * sets destination address prefix. CIDR or source IP range. Asterix â€œ*â€" severity="W" />

702		<issue line="566" char="63" reason="This character may get silently deleted by one or more browsers." evidence=" * traffic is allowed or denied. Possible values are â€œAllowâ€ and" severity="W" />

703		<issue line="567" char="12" reason="This character may get silently deleted by one or more browsers." evidence=" * â€œDenyâ€. Possible values for this property include: &apos;Allow&apos;, &apos;Deny&apos;." severity="W" />

704	</file>

705</jslint>

706

707npm ERR! Test failed.  See above for more details.

708npm ERR! not ok code 0

```

amarzavery commented:
@DeepakRajendranMsft - Please write tests for the generated code. This was not done previously for hyak generated code. However, we expect partner teams to write Basic CRUD tests for the generated code. This has multiple benefits:
- Partner team gets a first hand user experience of the generated code
- Tests can serve as samples for our customers
- If there are any bugs in the generated code they can be caught early on

Please follow the guide on writing tests in the documentation folder of AutoRest branch. You can also take a look at storage or resource management tests as an example.

amarzavery commented on lib/services/networkManagement2/generate.cmd:
```diff
@@ -0,0 +1,17 @@
+    set specFile="%1"
+)
+set repoRoot=%~dp0..\..\..\
+set generateFolder=%~dp0Lib
```
"l" of lib should be lowercase.

DeepakRajendranMsft replied:
fixed

amarzavery commented:
In package.json
- the version should be updated to 0.12.0 as this would be breaking changes to the generated code
- "main": "./lib/network.js", should be modified to "main": "./lib/networkManagementClient.js"
- "dependencies" section should be updated to the following: `{ "ms-rest" : "^1.2.0", "ms-rest-azure" : "^1.2.0" }`

romangromov commented:
@DeepakRajendranMsft the PublicIpAddress `deleteMethod` is broken:

in file `Lib\operations\publicIPAddresses.js`, line **119**:

`119: PublicIPAddresses.prototype.beginDelete = function () ...` must be 

`119: PublicIPAddresses.prototype.beginDeleteMethod = function () ...` instead

amarzavery commented on lib/services/networkManagement2/generate.cmd:
```diff
@@ -0,0 +1,17 @@
+::
+
+@echo off
+set autoRestVersion=0.13.0-Nightly20151115
```
`set autoRestVersion=0.13.0-Nightly20151126` This will resolve the issue of reserved words in long running operation names as pointed out by Roman in email conversation.

DeepakRajendranMsft replied:
fixed

romangromov commented:
@DeepakRajendranMsft can you please resolve reported issues?

amarzavery commented:
@DeepakRajendranMsft - The PR for fixing quotes in the swagger spec has been merged. https://github.com/Azure/azure-rest-api-specs/pull/88

DeepakRajendranMsft commented:
@amarzavery @romangromov just back from vacation. Will fix the reported issues.

romangromov commented:
Also, @DeepakRajendranMsft `LocalNetworkGateways.listAll` method is missed from this SDK

DeepakRajendranMsft commented:
@romangromov : Subnets never had a list all method. Looks like LocalNetworkGateways didnt have them either (looks like missing functionality here) @Nilambari  to comment on this.

@amarzavery: Thanks for your detailed look at this. I have addressed all your comments except tests. There is a parallel thread going on with your team about testing other languages. Lets deal with this there. (Its very hard and time consuming for service teams to write tests in all languages.)

DeepakRajendranMsft commented:
I have created a new PR @ https://github.com/Azure/azure-sdk-for-node/pull/1565
